File: Data Structures & Algorithms/subsets-ii/submission-25.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsetsWithDup(nums) {
        nums.sort((a, b) => a - b);
        const res = [[]];
        let prevIdx = 0;
        let idx = 0;
        for (let i = 0; i < nums.length; i++) {
            idx = (i >= 1 && nums[i] === nums[i - 1]) ? prevIdx : 0;
            prevIdx = res.length;
            for (let j = idx; j < prevIdx; j++) {
                const tmp = [...res[j]];
                tmp.push(nums[i]);
                res.push(tmp);
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-25.cs
public class Solution {
    private static List<List<int>> res = new List<List<int>>();

    public List<List<int>> CombinationSum2(int[] candidates, int target) {
        res.Clear();
        Array.Sort(candidates);

        dfs(0, new List<int>(), 0, candidates, target);
        return res;
    }

    private void dfs(int idx, List<int> path, int cur, int[] candidates, int target) {
        if (cur == target) {
            res.Add(new List<int>(path));
            return;
        }
        for (int i = idx; i < candidates.Length; i++) {
            if (i > idx && candidates[i] == candidates[i - 1]) {
                continue;
            }
            if (cur + candidates[i] > target) {
                break;
            }

            path.Add(candidates[i]);
            dfs(i + 1, path, cur + candidates[i], candidates, target);
            path.RemoveAt(path.Count - 1);
        }
    }
}